mmv1/third_party/terraform/services/bigquery/data_source_google_bigquery_tables.go

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package bigquery
+
+import ( 
+  "context" 
+  "fmt"
+  "log"
+  "time" 
+
+  bq "google.golang.org/api/bigquery/v2"
+
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+  "github.com/hashicorp/terraform-provider-google/google/tpgresource"
+  transport_tpg "github.com/hashicorp/terraform-provider-google/google/transport"
+)
+
+func DataSourceGoogleBigQueryTables() *schema.Resource {
+
+  dsSchema := map[string]*schema.Schema{
+    "dataset_id": {
+      Type:        schema.TypeString,
+      Required:    true,
+      Description: "The ID of the dataset containing the tables.",
+    },
+    "project": {
+      Type:        schema.TypeString,
+      Optional:    true,
+      Description: "The ID of the project in which the dataset is located. If it is not provided, the provider project is used.",
+    }, 
+    "tables": { 
+      Type:        schema.TypeList,
+      Computed:    true,
+      Elem:        &schema.Schema{Type: schema.TypeString},
+      Description: "A list of table names in the dataset.",
+    }, 
```

Comment:
Could the data source contain more information about the tables beyond their names?

One pattern we use in the provider currently is to take the schema from a resource and re-use it when implementing data sources, so that data sources' schemas don't repeat information unnecessarily. [Here's an example](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/kms/data_source_google_kms_crypto_keys.go#L15-L17) where the schema is obtained from a resource (you would use [ResourceBigQueryTable](https://github.com/hashicorp/terraform-provider-google/blob/7a4da3410a94796cd12d05f12a77b793888c7e8f/google/services/bigquery/resource_bigquery_table.go#L436) instead) and is altered to make sure any Required fields are changed to optional.

There might be some complications with this but wanted to raise the idea in case you weren't aware this was possible!

Reply:
Hey Sarah! It definitely can. This is honestly one of my first OS contributions so I'm gonna appreciate your advice here:
1. The attribute/schema structure designed here (including only table names) was purely following the ask from the issue itself (which I did not create, but do have an immediate use case for). I'm not sure what exactly to include to enrich this datasource, and would appreciate your thoughts here. Happy to go back to the drawing board to make the refinements then.
2. I'm also not able to review the error logs here: https://github.com/GoogleCloudPlatform/magic-modules/pull/11552#issuecomment-2313168435, so not sure how to go about fixing any issues. Any help here would also be much appreciated.

Reply:
@anubhavsharma515 No worries, thanks for coming and contributing!

> I'm not sure what exactly to include to enrich this datasource

Instead of the `tables` field on this data source being a list of names it could instead be a list of objects that have multiple fields describing each field. You range over the `Tables` list in your code, and at that point you can access [any of the values in this struct](https://pkg.go.dev/google.golang.org/api/bigquery/v2#TableListTables) for each table. Your code would need to map the fields in the client library's struct(s) to the data source's schema, e.g. [this code](https://github.com/GoogleCloudPlatform/magic-modules/blob/049127c69be365f0538dc9877b4ffdb27d8083f7/mmv1/third_party/terraform/services/kms/data_source_google_kms_crypto_key_versions.go#L233-L255) shows how you'd make a map[string]interface{} per table where the map keys are field names, and collect those maps in an []interface{}.

However there's no guarantee that the data returned from the client library when listing resources is as comprehensive as if you were retrieving the details of a single table (hence why re-using the schema from `google_bigquery_table` might not be 100% appropriate). In that case it's a choice of what makes sense to implement vs not. **@/wj-chen's suggestion to just implement names to start with is be a good idea here!** However you'd still need to update the PR so that the list is a list of objects instead of a list of strings.



>I'm also not able to review the error logs here

If you're unable to run the tests locally then us reviewers can give feedback about what needs fixing


---

>Is it a convention to return the full resource in a data source list? If we only support names but want to later on expand to the full resource, would that be considered a breaking change?

@wj-chen I don't think there are very established conventions here as there aren't many plural data sources! In the past I've tried to ensure as much data is present as possible for each item listed, but that may be unnecessary. As a counter-example [google_storage_buckets](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/storage/data_source_google_storage_buckets.go) includes only a [subset of the possible fields](https://github.com/GoogleCloudPlatform/magic-modules/blob/049127c69be365f0538dc9877b4ffdb27d8083f7/mmv1/third_party/terraform/services/storage/data_source_google_storage_buckets.go#L30-L50). I think implementing only the name field to start is valid, but it would require the swap from list of strings to list of maps.

Reply:
I think to avoid potentially having to update the list from just name strings to table objects, which I believe would be a breaking change for users, it might be worth exposing table objects from the get-go. On the other hand, the BigQuery Table schema is fairly involved, and I haven't done any work with data sources in the past myself to know if there might be complications mapping the schemas. I will let @anubhavsharma515 explore a bit.

If it becomes too much of a headache, we can also maybe just name this data source `google_bigquery_table_names` and scope it down to just name strings. What do you both think?

Reply:
Hey folks! I appreciate the patient feedback here.

I'm not apposed to enriching the legwork of investigating what makes sense to go in here at all. However, and this may come with a hint of selfishness, I believe that the availability of this data source with even the table names is of immense value for several use-cases. That being said, I am fully aligned on the concerns with having this attribute be a `list` - especially since future enrichment will likely be cumbersome as a result of.

For the time-being, I believe a fair implementation involves incorporating @SarahFrench's recommendation to make the attribute a map. @wj-chen would you say it's fair to leave the name of this data source as `google_bigquery_tables` following this change?

Reply:
I misunderstood the suggestion earlier, per my current understanding, yes let's keep the data source name `google_bigquery_tables`. And if my current understanding is correct, `tables` will still be a list, but now a list of maps where each map is a representation of a [`google_bigquery_table`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_table), in the form of key-value pairs of a `google_biguqery_table` attributes to their values. For now we want to return only the table names, so each map will have only one entry looking like `table_id`=`<the table name>`. In the future we can add more attributes. @SarahFrench Let me know if my understanding is correct.

Reply:
@wj-chen @SarahFrench my apologies, I completely misunderstood the structure of the changes requested. As advised, the following structure for `tables` attribute has been implemented:

```
tables = [
  {
    "table_id" = "<string>",
    "labels" = {
        <key/value pair>....
    },
  },
  ....
]
```